### ISSUE TITLE
Fix S2259 FP/FN: 'variable' is null on at least one execution path.

### ISSUE THREAD

**Issue by marco-carvalho** [SonarSource/sonar-dotnet#8001]
### Description

I'm checking for not null before, and Sonarqube is telling me is null on at least one execution path.

### Repro steps

```cs
GameLogger.Verify(_asset is not null && Stages.ContainsKey(_asset.Guid));
```

### Expected behavior

Not mark it with S2259.

### Actual behavior

Sonarqube is marking it with S2259.

![image](https://github.com/SonarSource/sonar-dotnet/assets/15821724/bed43deb-8b3f-4104-8864-d779142241e0)

**Comment by gregory-paidis-sonarsource** [comment]
Hey @marco-carvalho , 

There has been some work in this rule recently and this seems to work in the latest release of sonar-dotnet. 
I tried it locally with this snippet:

```csharp
class Human { public int Age = 42; }

bool Method(Human arg) => arg is not null && Check(arg.Age); // Compliant as expected

bool Check(int arg) => true;
```

Please let me know if this trivial example is equivalent to your case.
If it is one-to-one, all you have to do is wait for the next release of SQ :)
If your scenario is more complex or different, please provide me a working snippet to investigate it more thoroughly.

**Comment by marco-carvalho** [comment]
Hi @gregory-paidis-sonarsource ,

https://sonarcloud.io/project/issues?resolved=false&rules=csharpsquid%3AS2259&id=marco-carvalho_murder&open=AYqOaY_zMBeWHRncrANt

**Comment by gregory-paidis-sonarsource** [comment]
Hey again Marco,

Thanks for your quick reply!
I see what the problem is now. You are using automatic analysis, which has less precision than normal, CI-based analysis.

This is also why you get a warning on the top-right:

![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/c1e92645-6941-4d86-b37f-779496925554)

If you get a lot of noise like this, I would suggest you migrate your analysis to a pipeline, to enjoy the benefits of higher accuracy and less noise.

We are working continuously on improving automatic analysis as well, but it is still a fairly new product, and for now there is a tradeoff between ease of use and quality of diagnostics.

Thanks again,
Greg

**Comment by marco-carvalho** [comment]
https://docs.sonarcloud.io/advanced-setup/ci-based-analysis/overview-of-integrated-cis/

I think this is the correct link.

**Comment by gregory-paidis-sonarsource** [comment]
Thanks for catching this!
I informed Sonarcloud team via our internal channels :)
